datafusion/optimizer/src/extract_equijoin_predicate.rs

```diff
@@ -67,67 +67,145 @@ impl OptimizerRule for ExtractEquijoinPredicate {
+                join_constraint: join.join_constraint,
+                schema: join.schema,
+                null_equals_null,
+            }))
+        }
+    }
+    if join.filter.is_none() {
+        return Ok(Transformed::no(LogicalPlan::Join(join)));
+    }
+    let expr = join.filter.as_ref().unwrap();
+
+    let left_schema = join.left.schema();
+    let right_schema = join.right.schema();
+
+    if join.on.is_empty() {
+        let (eq_predicates, null_equals_null, non_eq_expr) =
+            split_eq_and_noneq_join_predicate(expr, left_schema, right_schema)?;
+        Ok(update_join_predicate(
+            join,
+            eq_predicates,
+            non_eq_expr,
+            null_equals_null,
+        ))
+    } else if join.null_equals_null {
+        let (eq_predicates, non_eq_expr) =
+            split_eq_and_noneq_join_predicate_nulls_eq(expr, left_schema, right_schema)?;
+
+        Ok(update_join_predicate(
+            join,
+            eq_predicates,
+            non_eq_expr,
+            true,
+        ))
+    } else {
+        let (eq_predicates, non_eq_expr) =
+            split_eq_and_noneq_join_predicate_nulls_not_eq(
+                expr,
+                left_schema,
+                right_schema,
+            )?;
+
+        Ok(update_join_predicate(
+            join,
+            eq_predicates,
+            non_eq_expr,
+            false,
+        ))
+    }
+}
+
 fn split_eq_and_noneq_join_predicate(
-    filter: Expr,
+    filter: &Expr,
+    left_schema: &DFSchema,
+    right_schema: &DFSchema,
+) -> Result<(Vec<EquijoinPredicate>, bool, Option<Expr>)> {
+    let (eq, noneq) = split_eq_and_noneq_join_predicate_nulls_not_eq(
+        filter,
+        left_schema,
+        right_schema,
+    )?;
+    if !eq.is_empty() {
+        Ok((eq, false, noneq))
+    } else {
+        let (eq, noneq) = split_eq_and_noneq_join_predicate_nulls_eq(
+            filter,
+            left_schema,
+            right_schema,
+        )?;
+        Ok((eq, true, noneq))
+    }
+}
+
+fn split_eq_and_noneq_join_predicate_nulls_eq(
+    filter: &Expr,
+    left_schema: &DFSchema,
+    right_schema: &DFSchema,
+) -> Result<(Vec<EquijoinPredicate>, Option<Expr>)> {
+    split_eq_and_noneq_join_predicate_impl(
+        filter,
+        left_schema,
+        right_schema,
+        Operator::IsNotDistinctFrom,
+    )
+}
+
+fn split_eq_and_noneq_join_predicate_nulls_not_eq(
+    filter: &Expr,
+    left_schema: &DFSchema,
+    right_schema: &DFSchema,
+) -> Result<(Vec<EquijoinPredicate>, Option<Expr>)> {
+    split_eq_and_noneq_join_predicate_impl(
+        filter,
+        left_schema,
+        right_schema,
+        Operator::Eq,
+    )
+}
+
+fn split_eq_and_noneq_join_predicate_impl(
+    filter: &Expr,
     left_schema: &DFSchema,
     right_schema: &DFSchema,
+    eq_op: Operator,
 ) -> Result<(Vec<EquijoinPredicate>, Option<Expr>)> {
-    let exprs = split_conjunction_owned(filter);
+    let exprs = split_conjunction(filter);
```

Comment:
Just a note: this part (along with cloning exprs) in `else`s of this function was changed in https://github.com/apache/datafusion/pull/10165 as a part of reducing expr cloning in optimizer initiative, so, probably, in this place it's preferred to avoid cloning, unless it's required.